src/jdk.compiler/share/classes/com/sun/tools/javac/comp/TransPatterns.java

```diff
@@ -640,6 +675,167 @@ JCNewClass makeNewClass(Type ctype, List<JCExpression> args) {
+                        } else {
+                            newGuard = null;
+                            instanceofCheck = (JCInstanceOf) accummulatedFirstLabel.guard;
+                        }
+                        JCBindingPattern binding = (JCBindingPattern) instanceofCheck.pattern;
+                        hasUnconditional =
+                                instanceofCheck.allowNull &&
+                                types.isSubtype(commonNestedExpression.type,
+                                                types.boxedTypeOrType(types.erasure(binding.type))) &&
+                                accList.tail.isEmpty();
+                        List<JCCaseLabel> newLabel;
+                        if (hasUnconditional) {
+                            newLabel = List.of(make.ConstantCaseLabel(makeNull()),
+                                               make.DefaultCaseLabel());
+                        } else {
+                            newLabel = List.of(make.PatternCaseLabel(binding, newGuard));
+                        }
+                        nestedCases.add(make.Case(CaseKind.STATEMENT, newLabel, accummulated.stats, null));
+                    }
+                    if (!hasUnconditional) {
+                        JCContinue continueSwitch = make.Continue(null);
+                        continueSwitch.target = currentSwitch;
+                        nestedCases.add(make.Case(CaseKind.STATEMENT,
+                                                  List.of(make.ConstantCaseLabel(makeNull()),
+                                                          make.DefaultCaseLabel()),
+                                                  List.of(continueSwitch),
+                                                  null));
+                    }
+                    JCSwitch newSwitch = make.Switch(commonNestedExpression, nestedCases.toList());
+                    newSwitch.patternSwitch = true;
+                    JCPatternCaseLabel leadingTest =
+                            (JCPatternCaseLabel) accummulator.first().labels.head;
+                    leadingTest.guard = null;
+                    result.add(make.Case(CaseKind.STATEMENT,
+                                         List.of(leadingTest),
+                                         List.of(newSwitch),
+                                         null));
+                } else {
+                    result.addAll(accummulator);
+                }
+                accummulator.clear();
+        };
+
+        VarSymbol commonBinding = null;
+        JCExpression commonNestedExpression = null;
+        VarSymbol commonNestedBinding = null;
+
+        for (List<JCCase> c = inputCases; c.nonEmpty(); c = c.tail) {
+            VarSymbol currentBinding;
+            JCExpression currentNestedExpression;
+            VarSymbol currentNestedBinding;
+
+            if (c.head.labels.size() == 1 &&
+                c.head.labels.head instanceof JCPatternCaseLabel patternLabel &&
```

Comment:
nit:
```
               (c.head.labels.size() == 1 &&
                c.head.labels.head instanceof JCPatternCaseLabel patternLabel)
```
seems to be a common condition that is evaluated twice in this `if` and in the one in the `else if` below, consider evaluating it only once